st2api/st2api/controllers/exp/inquiries.py

```diff
@@ -0,0 +1,273 @@
+        inquiry = self._get_one_by_id(
+            id=inquiry_id,
+            requester_user=requester_user,
+            permission_type=PermissionType.INQUIRY_RESPOND
+        )
+
+        if not self._inquiry_sanity_check(inquiry):
+            return
+
+        if not requester_user:
+            requester_user = UserDB(cfg.CONF.system_user.user)
+
+        # Determine permission of this user to respond to this Inquiry
+        if not self._can_respond(inquiry, requester_user):
+            abort(
+                http_client.FORBIDDEN,
+                'Insufficient permission to respond based on Inquiry parameters.'
+            )
+            return
+
+        # Validate the body of the response against the schema parameter for this inquiry
+        schema = inquiry.schema
+        LOG.debug("Validating inquiry response: %s against schema: %s" %
+                  (response_data.response, schema))
+        try:
+            util_schema.validate(instance=response_data.response, schema=schema,
+                                 cls=util_schema.CustomValidator, use_default=True,
+                                 allow_default_none=True)
+        except Exception as e:
+            LOG.debug("Failed to validate response data against provided schema: %s" % e.message)
+            abort(http_client.BAD_REQUEST, 'Response did not pass schema validation.')
+            return
+
+        # Update inquiry for completion
+        new_result = copy.deepcopy(inquiry.result)
+        new_result["response"] = response_data.response
+        liveaction_db = self._mark_inquiry_complete(
+            inquiry.liveaction.get('id'),
+            new_result
+        )
+
+        # We only want to request a workflow resume if this has a parent
+        if liveaction_db.context.get("parent"):
+
+            # Request that root workflow resumes
+            root_liveaction = action_service.get_root_liveaction(liveaction_db)
+            action_service.request_resume(
+                root_liveaction,
+                requester_user
+            )
+
+        return {
+            "id": inquiry_id,
+            "response": response_data.response
+        }
+
+    def _inquiry_sanity_check(self, inquiry_candidate):
+        """Sanity checks for ensuring that a retrieved execution is indeed an Inquiry
+
+        It must use the "inquirer" runner, and it must currently be in a "pending" status
+
+        :param inquiry_candidate: The inquiry to check
+
+        :rtype: bool - True if a valid Inquiry. False if not.
+        """
+
+        if inquiry_candidate.runner.get('runner_module') != "inquirer":
+            abort(http_client.BAD_REQUEST, '%s is not an Inquiry.' % inquiry_candidate.id)
+            return False
+
+        if inquiry_candidate.status != "pending":
+            abort(http_client.BAD_REQUEST,
+                  'Inquiry %s has already been responded to' % inquiry_candidate.id)
+            return False
```

Comment:
After abort, does this method still return False?

Reply:
It doesn't; I only included this to be a bit more explicit. Elsewhere in this controller and in other controllers, a return statement is included after an `abort()` call, presumably for the same reason. So I was just trying to keep it consistent.

Reply:
So I would make this function either return True or False. If we're aborting on check failure, then by not erring/aborting, this mean check succeed. I know you're trying to keep it consistent w/ other functions. My guess is those functions are not implemented correctly if that's the case.

Reply:
Okay. Another reason this was done was to raise different error messages depending on the different failure scenarios. I could move the `abort()` call back outside the function so the function always returns True/False, and also return the error message if any, so I can still provide feedback to the user.

Reply:
Done in 577d019f0